activerecord/lib/active_record/attribute_methods/write.rb

```diff
@@ -36,8 +36,7 @@ def write_attribute(attr_name, value)
         end
 
         name = self.class.primary_key if name == "id".freeze && self.class.primary_key
-        @attributes.write_from_user(name, value)
-        value
+        _write_attribute(name, value)
       end
 
       def raw_write_attribute(attr_name, value) # :nodoc:
```

Comment:
Seems slightly unfortunate (read: confusing) to have both `raw_write_attribute` and `_write_attribute`, but I don't have any concrete suggestions for avoiding it.